### PR TITLE
Feature/fonts improvement

### PR DESCRIPTION
**Pull request by luka-nextcloud**
* Resolves: #3139
* Target version: main

### Summary

- Add repair step to install default fonts on installation
- Add command richdocuments:install-fonts  to install default fonts

### TODO

- Font files get shipped as part of richdocuments in an extra assets/fonts/ directory
- New fonts shall be copied from the assets/fonts folder
- Add a repair job to add new fonts once to the installation
- Add an occ command richdocuments:install-fonts to install shipped fonts again if manually removed as the admin
- Add cypress test that ensures the fonts are listed in the admin settings

### CODE REVIEW

**Comment by kesselb** [comment]
Tip: If composer dump-autoload does not update the autoloader then `cd composer`. I just ran into the same failing workflow today and was super confused by it :see_no_evil:

File: lib/Command/InstallDefaultFonts.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class InstallDefaultFonts extends Command {
+	private FontService $fontService;
+
+	public function __construct(FontService $fontService) {
+		parent::__construct();
+		$this->fontService = $fontService;
+	}
```
Comment (juliushaertl) [inline lib/Command/InstallDefaultFonts.php:16]: We can make use of PHP 8 constructor property promotion:

https://php.watch/versions/8.0/constructor-property-promotion

```suggestion
	public function __construct(private FontService $fontService) {
		parent::__construct();
	}
```

File: lib/Migration/InstallDefaultFonts.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+use OCP\Migration\IRepairStep;
+
+class InstallDefaultFonts implements IRepairStep {
+	private IConfig $config;
+	private FontService $fontService;
+
+	public function __construct(IConfig $config, FontService $fontService) {
+		$this->config = $config;
+		$this->fontService = $fontService;
+	}
```
Comment (juliushaertl) [inline lib/Migration/InstallDefaultFonts.php:17]:
```suggestion
	public function __construct(private IConfig $config, private FontService $fontService) {
	}
```

File: lib/Service/FontService.php [hunk 3]
```diff
@@ -242,4 +243,41 @@ private function generateFontOverview(ISimpleFile $fontFile): void {
...
+				continue;
+			}
+
+			$filePath = $dirPath . '/' . $fileName;
+
+			if (!is_file($filePath)) {
+				continue;
+			}
+
+			$fileContent = file_get_contents($filePath);
```
Comment (juliushaertl) [inline lib/Service/FontService.php:274]: While this is fine for this case a general note that using file_get_contents would read the full file content into memory which might cause problems with larger files. Better approach would be to use fopen instead and pass a file pointer instead which is also supported by the `newFile` method that is called within `uploadFontFile`

**Review by juliushaertl — COMMENTED** [review]
Haven't tested yet, but looks good. Only a few minor comments inline.

File: lib/Command/InstallDefaultFonts.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace OCA\Richdocuments\Command;
+
+use OCA\Richdocuments\Service\FontService;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class InstallDefaultFonts extends Command {
```
Comment (juliushaertl) [inline lib/Command/InstallDefaultFonts.php:10]: Ah we still don't have a public API for commands 🙈 We can add the required classes as a stub to tests/stub.phpstub

**Comment by juliushaertl** [comment]
@luka-nextcloud Please make sure to provide some context like a short pr description and the link to the issue

**Comment by juliushaertl** [comment]
> Tip: If composer dump-autoload does not update the autoloader then cd composer. I just ran into the same failing workflow today and was super confused by it

Yeah we should probably document that better. I went for a separate composer.json for the autoloader only to avoid dev dependencies getting autoloaded, but not the most straight forward I agree

**Review by juliushaertl — APPROVED** [review]

**Comment by juliushaertl** [comment]
Pushed a commit adding the fonts and licenses through a build script.

**Pull request merged**